Faults

mrevang opened the pull request:
### Change Scope

This proposal introduces a list of faults under a component's healthz container.

A fault indicates that a component has recently or is currently experiencing a problem as expressed by a symptom. These symptoms have underlying conditions that are platform-specific. The fault should not be deleted as soon as the underlying condition is no longer asserted to avoid excessive fault creation and deletion. Deletion of the fault should occur after a suitable period outside of the triggering condition

Nested under each fault is a list of actions recommended to remediate the fault condition.

```
        +--rw oc-platform-healthz:healthz
        |  +--rw oc-platform-healthz-fault:faults <<<<<<<<<<<<<<<<<<<<<<<
        |     +--ro oc-platform-healthz-fault:fault* [symptom]
        |        +--ro oc-platform-healthz-fault:symptom         -> ../state/symptom
        |        +--ro oc-platform-healthz-fault:state
        |        |  +--ro oc-platform-healthz-fault:symptom?               identityref
        |        |  +--ro oc-platform-healthz-fault:origin-time?           oc-types:timeticks64
        |        |  +--ro oc-platform-healthz-fault:last-detection-time?   oc-types:timeticks64
        |        |  +--ro oc-platform-healthz-fault:description?           string
        |        |  +--ro oc-platform-healthz-fault:status?                enumeration
        |        |  +--ro oc-platform-healthz-fault:counters
        |        |     +--ro oc-platform-healthz-fault:occurrences?   oc-yang:counter64
        |        +--ro oc-platform-healthz-fault:remediations
        |           +--ro oc-platform-healthz-fault:remediation* [index]
        |              +--ro oc-platform-healthz-fault:index    -> ../state/index
        |              +--ro oc-platform-healthz-fault:state
        |                 +--ro oc-platform-healthz-fault:index?    uint64
        |                 +--ro oc-platform-healthz-fault:action?   identityref
        |                 +--ro oc-platform-healthz-fault:target?   -> /oc-platform:components/component/name

```

This change is backwards compatible.

### Platform Implementations

We are working with a few vendors offline to support this model directly. Here are some publicly available references that show devices reporting faults:
* https://www.cisco.com/c/en/us/td/docs/ios-xml/ios/bsm/configuration/15-2mt/bsm-troubleshooting.html ?
* https://www.juniper.net/documentation/us/en/software/junos/cli-reference/topics/ref/command/show-system-errors.html

mrevang commented on release/models/platform/openconfig-platform-healthz-fault.yang:
```diff
@@ -0,0 +1,301 @@
+      "Base identity for fault symptoms";
+  }
+
+  identity ACTION_NON_IMPACTING_RESEAT {
```
"Reseat component"

mrevang commented on release/models/platform/openconfig-platform-healthz-fault.yang:
```diff
@@ -0,0 +1,301 @@
+  identity ACTION_FACTORY_RESET {
+    base ACTION_BASE;
+    description
+      "Restores a device to its original factory settings.";
```
device->component

mrevang commented on release/models/platform/openconfig-platform-healthz-fault.yang:
```diff
@@ -0,0 +1,301 @@
+  identity ACTION_POWER_CYCLE {
+    base ACTION_BASE;
+    description
+      "The process of turning a device off completely and then turning it back
```
device->component

mrevang commented on release/models/platform/openconfig-platform-healthz-fault.yang:
```diff
@@ -0,0 +1,301 @@
+      type oc-types:timeticks64;
+      description
+        "The time at which the fault's underlying condition was last sampled
+        and confirmed asserted.";
```
Add an MD file to provide examples.

mrevang commented on release/models/platform/openconfig-platform-healthz-fault.yang:
```diff
@@ -0,0 +1,301 @@
+          leaf index {
+            type uint64;
+            description
+            "This index indicates the order in which the actions should be
```
The order should progress from least to most disruptive.

dplore commented:
/gcbrun

dplore commented:
/gcbrun

ghananigans commented on release/models/platform/openconfig-platform-healthz-fault.yang:
```diff
@@ -0,0 +1,293 @@
+            type identityref {
+              base ACTION_BASE;
+            }
+          }
```
Should this only be one action? Are there scenarios when multiple actions may be performed to address the fault? e.g. `ACTION_WARM_REBOOT` _or_ `ACTION_COLD_REBOOT`

Alternatively, is there any ordering between actions where performing action A also covers remediations performed by action B?

ghananigans commented on release/models/platform/openconfig-platform-healthz-fault.yang:
```diff
@@ -0,0 +1,293 @@
+    leaf origin-time {
+      type oc-types:timeticks64;
+      description
+        "The time at which the fault first became active.
```
Is this time since boot-time?

ghananigans commented:
Apologies, i didn't realize this was a draft. I only noticed that `public-approvers` was added as a reviewer.

mrevang commented:
> Discussed in the OC Operators meeting on August 26th:
> 
> Healthz is _meant_ to simplify the health of the device to remove some of the if / then conditionals on actions that the operator needs to worry about. In its current form, I could see cases where this would 'encourage' a status of "unhealthy" + "minor non-actionable symptom", which seems antithetical to the healthz container.
> 
> The authors will see about changing the language / allowable severities in the container to avoid these types of situations.

I've introduced a new identityref for severity instead of reusing the identityref from alarms. I've also updated the language to indicate that faults should only be reported if the underlying condition rises to the existing "healthz-unhealthy" bar.

dplore commented on release/models/platform/openconfig-platform-healthz-fault.yang:
```diff
@@ -0,0 +1,318 @@
+    base OPENCONFIG_FAULT_SEVERITY;
+    description
+      "Indicates that a fault does not require urgent remediation.";
+  }
```
My thinking is that it is very likely that the network management system will need to determine this according to the an operator's policies for running their network.  So my bias is this should not be included in the model.

If there are some operational scenarios that could be  quantified / measured that can be used to know the urgency then maybe these are justified.  I think we'd need to see some examples and see if there is some pattern that can be measured and modeled.

dplore commented:
Reviewed in Sep 9, 2025 OC operators meeting.  Main feedback is the need for clearer definition of severities and how these are set, or possibly removing severity from the model.

dplore reviewed (changes requested):
Some comments on severity

dplore commented:
/gcbrun

mrevang commented on release/models/platform/openconfig-platform-healthz-fault.yang:
```diff
@@ -0,0 +1,293 @@
+      description
+        "Enclosing container for a fault's recommended remediations.";
+
+      list remediation {
```
Updated the remediation description to help clarify the intention/expectation around its use.

dplore replied:
My bias given your workflow description of "try this then that" also makes me think an ordered list is probably more explicit and better.  The operator can still chose to execute in a different order or skip repair actions.

dplore commented:
Moved back to ready to discuss for the next OC operators meeting on Sep 30, 2025

ElodinLaarz commented:
Discussed in OC Operators Meet on Oct 07: Once any remaining changes are included here, we can move it to last-call and get any further input from folks. (Just ping me or @dplore to move it to last-call once it's ready.)

dplore commented:
/gcbrun

earies commented on release/models/platform/openconfig-platform-healthz-fault.yang:
```diff
@@ -0,0 +1,311 @@
+
+    leaf last-detection-time {
+      type oc-types:timeticks64;
+      description
```
Per `oc-types:timeticks64` definition, this leaf must define the epoch referenced (Can see other uses for examples)

earies commented on release/models/platform/openconfig-platform-healthz-fault.yang:
```diff
@@ -0,0 +1,311 @@
+    }
+
+    leaf occurrences {
+      type uint64;
```
`oc-yang:counter64`?  worthy of a counters sub-container?

mrevang replied:
Done.

mrevang commented:
Thanks for your comments and thoughts. I appreciate the feedback. I will try to address some of your points inline below.

> A few questions/comments:
> 
> Reading through this proposal, the desire is well understood.
> 
> However, what I would like to see is when proposals as such come across that encode what are likely to be new, altered or similar behavior, that we start with concrete problem statements, use-cases and examples including example instance-data payloads according to the modeling.
> 
> While there are vendor implementation references, those tend to merely be similarities and/or may not apply to implementation targets for such feature-sets.

* Problem statement: Alarms provide a way to express that the device is in some kind of distress. However, alarms lack component attribution or a suggested response to the alarm. This requires operators to build up this knowledge over time. The ambition of this feature is that vendors have this context (via FMEA) and are able to provide this additional context along with the distress signal. 
* Use-case: Operators will use this information to mitigate fault conditions that affect operation. I can provide concrete fault examples in a later update.

> 
> That aside:
> 
> * Components modeling decided to group in software components some time back but the verbiage and constructs used within here lean towards hardware only.  Since there are no gating statements, the possibilities are wide atm.  What if any thoughts are there around how this should be applied or handled with software components?  It is fine to say it's not but this lack of clarity or accompanying documents leaves the industry lost and diverged (not unique to this proposal but want to point this out)

This proposal is definitely motivated by hardware component failures and has that bias. I don’t see a reason to prevent fault reporting on software components should they rise to healthz’s “UNHEALTHY” standard.

> * TMK, we have not yet discussed fault handling yet (much like alarms).  Sure, we can reference prior art to similar approaches but we have a tree where items pop up, remidiation actions may assist but are these to be cleared automatically by the system?  Does a user have to ack/clear?  If so, how?

There is no expectation that the operator should ack/clear the event. If the underlying fault condition is no longer asserted, after a vendor-defined cool down period (in the interest of fault dampening), the fault is expected to be removed by the device.

> * How about fault categories or severities?

Severity was in the early proposal, but was ultimately removed. I believe the consensus was that if “component is no longer able to perform the function expected of it and requires remediation/intervention” that we’d always be in a “severe” place.

> * Historical and/or clear counts?  e.g. the fault is not active now but there was prior record

A historical counter was also discussed. Many of the remediations (reboot, reimage, ect) would require this history to persist across these actions, which was deemed expensive to support. The prevailing thought was that the operator receiving these faults could monitor for recidivism and escalate as necessary.
 
> I see this is tagged for last call w/o any implementor comments yet. Can last call be extended post the next OC community call (and maybe a topic to bring to the table)?

I was too slow adding this topic to the agenda and the time was already claimed. We can try for the next meeting if needed.

I will prepare an accompanying document that captures some of these questions and design decisions for future guidance.

nkitchen-arista commented on release/models/platform/openconfig-platform-healthz-fault.yang:
```diff
@@ -0,0 +1,330 @@
+
+  grouping fault-remediations {
+    description
+      "Grouping containing a faults recommended remediations";
```
faults -> fault's?

nkitchen-arista commented on release/models/platform/openconfig-platform-healthz-fault.yang:
```diff
@@ -0,0 +1,330 @@
+      "Component is missing";
+  }
+
+  identity SYMPTOM_MISWIRED {
```
The state of being miswired seems more like a fault cause than the visible symptom of a fault.

mrevang commented on doc/component_fault_overview.md:
```diff
@@ -0,0 +1,189 @@
+
+This proposal introduces a list of faults under a component's healthz container.
+
+A fault indicates that a component has recently or is currently experiencing a problem as expressed by a symptom. These symptoms have underlying conditions that are platform-specific. The fault should not be deleted as soon as the underlying condition is no longer asserted to avoid excessive fault creation and deletion. Deletion of the fault should occur after a suitable period outside of the tri ...
```
I don't think so. Ultimately if the component is performing its intended function, the operator should be happy. Providing feedback about why a condition was alleviated (operator intervention or otherwise) seems low value IMO.

> is this period going to be vendor-defined?

Yes. The concern with providing a value, is that there isn't a universally appropriate one. It may change from device, component, or even symptom. Ultimately the bar for a fault is set pretty high, ie that it's not performing the function expected of it. One could expect that once it reaches that bar, it's not likely to recover. However, with threshold breaches (temp high), instantaneous values can oscillate in and out of that condition. This proposal would like to prevent fault oscillation.

mrevang replied:
While we don't have a "why", the `status` leaf does provide some feedback that the underlying condition is no longer satisfied. I'll add a note to the markdown doc to highlight this point.

mrevang commented on release/models/platform/openconfig-platform-healthz-fault.yang:
```diff
@@ -0,0 +1,324 @@
+        "Enclosing container for a component's active faults.";
+
+      list fault {
+        key "symptom";
```
I'd argue this is desirable because it bounds concurrent faults. SENSOR is a supported component type, so a component could be spun off for each sensor in your example, that increased granularity is still expressible.

If one or both of the sensors on your component was out of range to the point that it prevented the component from performing its function, the status would be active.

Occurrences are meant to track transitions into the active state. In the scenario above, if the component continues to experience loss of functionality due to a rotating roster of sub-components being out of range, that would be a single occurrence.

dplore commented:
/gcbrun
